perf(GlobalSearch): bulk insert queued items

18alantom opened the pull request:
Fixes #14660

Global search queue items are popped and inserted into the `__global_search`
table. This used to happen individually.

This PR bulk inserts them in chunks of 10,000.

18alantom committed:
perf(GlobalSearch): bulk insert queued items

Fixes #14660

ankush commented on frappe/utils/global_search.py:
```diff
@@ -373,11 +373,42 @@ def sync_global_search():
+
+	search_queue_items = get_search_queue_items_generator()
+
+	while values := tuple(islice(search_queue_items, 10_000)):
```
There's create batch utils somewhere in framework.

Also can we somehow pop from redis in batches too? 🤔 

(Need to ensure that we don't miss anything though)

ankush replied:
Another question: what happens when same document gets modified twice and is sent twice via same query. DBs might not like it 😄

18alantom replied:
`RPOP KEY LEN` can performed, will have to alter the wrapper though, should I do that?

18alantom replied:
Okay, nvm the `LEN` argument was added in 6.20 we (CI, FC) run an older version of redis which doesn't have that.

18alantom replied:
Also `utils.create_batch` won't work for this purpose cause while it asks for an `Iterable`, before yielding it slices the `Iterable`.

`Iterable` does not imply sliceable.

Misconstrued typing aside, all items will have to popped from Redis for `utils.create_batch`'s use to be justified.

18alantom committed:
refactor: add count to cache pop

18alantom committed:
refactor: use rpop count to pop in batches

- fix: dedupe items in a batch
- fix: reassign `query` cause it does not hold state

_fkn ORMs man 🙄_

18alantom committed:
refactor: move inner for to function

18alantom committed:
refactor: undo use of count in redis pop

count param in rpop and lpop was introduced in 6.20
at the time of this commit we are using an older
version that does not include it

18alantom committed:
fix(sync_values): use EXCLUDED for postgres

removing the second arg from `do_update` causes
EXCLUDE to be auto added

ankush commented:
@18alantom I am assuming it's ready for review. Draft PRs usually imply otherwise.

ankush commented on frappe/utils/global_search.py:
```diff
@@ -373,11 +373,57 @@ def sync_global_search():
+		item_dict = json.loads(item_json)
+		key = (item_dict["doctype"], item_dict["name"])
+
+		if key in values_dict:
+			del values_dict[key]
+
```
```suggestion
```

Not required? You're overriding the key in next line.

18alantom replied:
<img width="308" alt="Screenshot 2023-11-06 at 11 49 14" src="https://github.com/frappe/frappe/assets/29507195/7a107acd-db24-41ac-bf80-7f9fd70ee7d9">

If insertion order is unimportant then yeah, not required.

Considering the [specific use of `rpop`](https://github.com/frappe/frappe/pull/23064/files/3d7ba41583ffabdddef007a9a7274024ab305c96#diff-2df9e640d5d5f2f19851093522f542bbdceea96852667ee33042ad22caa95ac5L377) I thought it would be.


(probably should've added a comment)

ankush replied:
@18alantom we just need the last write to win, it doesn't have to actually be in order. So we are fine I think. 

Also, python ordereddict isn't really required too, dict by default is ordered too (enough for this use case) :smile:

ankush commented on frappe/utils/global_search.py:
```diff
@@ -373,11 +373,57 @@ def sync_global_search():
-		# rpop to follow FIFO
-		# Last one should override all previous contents of same document
-		value = json.loads(frappe.cache.rpop("global_search_queue").decode("utf-8"))
-		sync_value(value)
```
deprecate old unused functions with `@deprecated`. We usually remove them after some time has passed or before next major release.

18alantom replied:
Got it, although if you're referring to `sync_value` it's used elsewhere.

ankush commented on frappe/utils/global_search.py:
```diff
@@ -373,11 +373,57 @@ def sync_global_search():
+		sync_values(values)
+
+
+def get_deduped_search_item_values(items):
```
```suggestion
def _get_deduped_search_item_values(items):
```

If things are in `utils/*` prefer prefixing internal functions with `_`, so they can be removed/modified without worrying about side effects.

ankush commented on frappe/utils/global_search.py:
```diff
@@ -373,11 +373,57 @@ def sync_global_search():
+
+	values_dict = OrderedDict()
+	for item in items:
+		item_json = item.decode("utf-8")
```
It's kinda weird that we've configured Redis client to not auto-convert to strings :woozy_face: 

Some other day ig.

ankush committed:
refactor: misc changes

- make internal function explicit
- dont delete keys which is immediately being replaced.
- deprecate `sync_value`

ankush committed:
perf: don't check len before popping

rpop is atomic and yields None when nothing is found. No need to check length.

git-avc commented:
Maybe is related?
https://github.com/frappe/frappe/issues/23106
